### PR TITLE
Give visual feedback when reloading the schema

### LINKED ISSUE #379: Visual indicator when schema was reloaded
Visual feedback is one of the most important components to a great UX. When developing a GraphQL server, you're changing your schema quite often. This workflow is being steadily improved by allowing `CMD + R` to reload the schema (#180) and even automatic schema reloads based on filewatching via graphqlconfig (#331).

I suggest to introduce a nice visual animation (some sort of pulse & rotation - maybe even colored) when the schema has been reloaded.

image

@xavcz maybe you have some idea how we could design/implement this? 👨‍🎨

### PR DESCRIPTION
**Pull request by xavcz**
> Fixes #379.

Changes proposed in this pull request:
--- 

screencast of the change

- Simple animation on the reload icon when the schema is reloading (↔ `GraphQLEditor`'s `onReload` method is triggered)

- Two SVG elements instead of the `graphcool-styles` SVG icon: when not reloading, show the icon, when reloading, show a spinner

Questions
---
> When developing a GraphQL server, you're changing your schema quite often. This workflow is being steadily improved by allowing CMD + R to reload the schema (#180) and even automatic schema reloads based on filewatching via graphqlconfig (#331).

Where are schema changes broadcasted? I'd love to listen to it and thus providing feedback but I couldn't find the code 🙈

### CODE REVIEW

File: packages/graphql-playground-react/src/components/Playground/GraphQLEditor.tsx [hunk 1]
```diff
@@ -757,14 +759,19 @@ export class GraphQLEditor extends React.PureComponent<
...
+        this.props.session.endpoint || this.props.endpoint,
+        this.convertHeaders(this.props.session.headers),
+      )
+      if (result) {
+        const { schema } = result
+        this.setState({ schema, isReloadingSchema: false })
+        this.renewStacks(schema)
+      }
+    } catch (e) {
+      this.setState({ isReloadingSchema: false })
```
Comment (xavcz) [inline packages/graphql-playground-react/src/components/Playground/GraphQLEditor.tsx]: What about showing some feedback when reloading the schema fails?

> How about displaying a little notification dot (similar how the Twitter feed does it) next to the reload icon which is there until you change/run a query.

Something like a red dot? 🔴 

With maybe a tooltip saying "invalid schema"?

Or reusing the "unreachable schema" state?

File: packages/graphql-playground-react/src/components/Playground/TopBar/ReloadIcon.tsx [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+  fill: ${iconColor};
+  transition: 0.1s linear all;
+
+  &:hover {
+    fill: ${iconColorHover};
+  }
+`
+
+const showWhenReloading = bool => props =>
+  props.isReloadingSchema ? Number(bool) : Number(!bool)
```
Comment (xavcz) [inline packages/graphql-playground-react/src/components/Playground/TopBar/ReloadIcon.tsx]: Weird workaround, I got a TS error I couldn't solve by doing:
```tsx
const Element: StyledComponentClass<any, any, any> = styled.xyz`
	/* ... */
   opacity: ${props => props.isReloadingSchema ? 1 : 0};
`
```

**Comment by Huvik** [comment]
Looks good 😊

File: packages/graphql-playground-react/src/components/Playground/TopBar/ReloadIcon.tsx (hunk 2 above)
Comment (timsuchanek) [reply packages/graphql-playground-react/src/components/Playground/TopBar/ReloadIcon.tsx]: I'm right now merging in a styled components PR and include a `withProps` function that will help with this.

You will be able to use the `withProps` function like this:
```ts
import { styled, withProps } from '../styled/styled-components'

export interface Props {
  center?: boolean
}

export const P = withProps<Props>()(styled.div)`
  display: inline-block;
  font-size: 14px;
  text-align: ${p => (p.center ? 'center' : 'left')};
`
```

I will notify you when it's ready!

**Comment by timsuchanek** [comment]
Right now, schema changes are not broadcasted, that's a separate issue. We could get that information in the electron app by listening to the local changes of a project.

File: packages/graphql-playground-react/src/components/Playground/GraphQLEditor.tsx (hunk 1 above)
Comment (timsuchanek) [reply packages/graphql-playground-react/src/components/Playground/GraphQLEditor.tsx]: Yes, we can reuse that state for this purpose.

**Comment by schickling** [comment]
OMG. This looks awesome!

File: packages/graphql-playground-react/src/components/Playground/TopBar/ReloadIcon.tsx (hunk 2 above)
Comment (xavcz) [reply packages/graphql-playground-react/src/components/Playground/TopBar/ReloadIcon.tsx]: Oh, excellent!

Comment (timsuchanek) [reply packages/graphql-playground-react/src/components/Playground/TopBar/ReloadIcon.tsx]: `withProps` just arrived, you can import it from here: https://github.com/graphcool/graphql-playground/blob/master/packages/graphql-playground-react/src/styled/styled.ts

**Comment by xavcz** [comment]
Types are good to go now!

This topbar thrives passion: I just reverted a simple animation on the error message (https://github.com/graphcool/graphql-playground/pull/516/commits/19b90a4e2d42b9c4448bc6e7fe8a2e3a89f40fa7) to avoid conflicts with #504. 

Let's work on the error message in another PR once both are merged?

**Comment by timsuchanek** [comment]
I had to do a small merge back from master, but it looks good so far, waiting for the build to succeed now.
What do you exactly mean with error message?
Yes, sounds like keeping it separate is a good idea.

**Comment by xavcz** [comment]
I meant coupling the `endpointUnreachable` state and this PR, like on this screencast below 😁

screencast

 🍡

**Comment by Weakky** [comment]
Wouldn't it be more consistent to just keep your work over mine now? The spinner you made is visible whenever the schema is reloading. When it fails to reach the server endpoint, the reason why there's a spinner is because it continues to try to fetch the schema until it succeeds. Therefore, it doesn't make a lot of sense to have two different spinners for the same underlying action?

**Comment by xavcz** [comment]
> When it fails to reach the server endpoint, the reason why there's a spinner is because it continues to try to fetch the schema until it succeeds

Nice, I didn't know that! Makes sense to only show one, you're right.

**Pull request merged**